_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,383 @@
+class Polygon
+end
+```
+And from here, we can already describe how we're going to solve our problem, by creating a `#new` instance of a `Polygon`, and calling `#perimeter` on it:
+```ruby
+Polygon.new(___).perimeter
+```
+> In Ruby, and in most OOP languages, a `class` describes a "template" for an object. An object is usually a collection of data, and a set of methods or functions that allow you to interact with that data or give it logic. We first describe the objects with a `class`, and create new instances of them with `Thing.new`.
+
+In the next sections, we'll work with one approach to solve this from the ground up, to figure out how to *make* a `Polygon` to get what we need, and explore some properties of Ruby along the way.
+
+## Objects
+
+### Point
+
+Consider any kind of object around you. A pen, a car, your computer - they all have properties, or *attributes*.  Ruby objects can be given *attributes* that let you describe and store data inside them. We do this with the `@var` syntax, like so, describing what defines a `Point` in 2D space:
+
+```ruby
+# A location in 2D space.
+class Point
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> `def` defines a new method, or an instruction on how to do something with an object. `initialize` is a special method that is called whenever you call `#new` on an object to make a new instance of one.
+
+We can now make a new `Point` like so:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.inspect #=> #<Point:0x2ae37f0 @x=1.5, @y=-3.0>
+```
+
+The attributes we've defined (`@x`, `@y`) can only be accessed inside of the class's own methods. We're going to need to use these attributes of a point in order to do our calculations. To do so, we can define methods that will return the values of our attributes:
+
+```ruby
+# A location in 2D space.
+class Point
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+
+  def x
+    @x
+  end
+
+  def y
+    @y
+  end
+end
+```
+Since this is such a common thing in Ruby, the language has shortcuts to quickly define such simple methods that reflect an objects attributes:
+
+- `attr_reader :var` - Creates a read-only method for attribute `@var`.
+- `attr_writer :var` - Creates a write-only method for attribute `@var`.
+- `attr_accessor :var` - Creates both kinds of methods to read and write to this attribute.
+
+So, our `Point` can be described simply as:
+```ruby
+# A location in 2D space.
+class Point
+  attr_reader :x
+
+  attr_reader :y
+
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> In our case, we will only need to ever read these attributes, so I just used `attr_reader`. `attr_accessor` would have also worked here.
+
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
```

Comment:
I think this is an appropriate place for this footer, as linking it earlier may distract someone from what I'm explaining if they open a link to the other site. Here, I've fully demonstrated attributes and getters/setters, and am about to move on to another topic ; this is a safe place to deviate from the lesson if someone wants to know more about them.